### PR TITLE
Added Upnp support, followed on from PR 422

### PR DESCRIPTION
**Pull request by johng**
The [PR](https://github.com/lightningnetwork/lnd/pull/422) was over 1 month old, so I had a go at implementing the comments on the PR.

Original request follows from the issue #185

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -198,10 +200,42 @@ func newServer(listenAddrs []string, chanDB *channeldb.DB, cc *chainControl,
...
-	// If external IP addresses have been specified, add those to the list
+	// Gather external IPs from config
+	externalIPs := cfg.ExternalIPs
+
+	if cfg.UpnpSupport {
+
+		// Connect to router
+		d, err := upnp.DiscoverCtx(context.Background())
+		if err != nil {
+			fmt.Printf("Upnp: Unable to discover router %v\n", err)
```
Comment (Roasbeef) [inline server.go]: This'll cause the daemon to halt atm if it can't find a UPnP enabled rotuer.

**Comment by johng** [comment]
@Roasbeef sorry for the delay on getting back to this. I've added NAT-PMP support, and it no longer kills the daemon on failure.  I've been unable to test the NAT-PMP support, I don't think my router is configured to accept the protocol.

**Comment by Roasbeef** [comment]
Modifying my set up a bit the `--upnp` option just causes startup to hang. Looking at the goroutine dump, this is where it's stuck at:
```
1 @ 0x102effc 0x1029faa 0x10295a7 0x10c31ce 0x10c324d 0x10c442a 0x117a85e 0x11991ca 0x1196fd2 0x17e4985 0x17e5879 0x180788b 0x180870a 0x1809ff2 0x1811343 0x186212b 0x18657a3 0x1838d22 0x183ae76 0x102eb5d 0x105da71
#	0x10295a6	internal/poll.runtime_pollWait+0x56													/usr/local/go/src/runtime/netpoll.go:173
#	0x10c31cd	internal/poll.(*pollDesc).wait+0xad													/usr/local/go/src/internal/poll/fd_poll_runtime.go:85
#	0x10c324c	internal/poll.(*pollDesc).waitRead+0x3c													/usr/local/go/src/internal/poll/fd_poll_runtime.go:90
#	0x10c4429	internal/poll.(*FD).ReadFrom+0x199													/usr/local/go/src/internal/poll/fd_unix.go:169
#	0x117a85d	net.(*netFD).readFrom+0x5d														/usr/local/go/src/net/fd_unix.go:208
#	0x11991c9	net.(*UDPConn).readFrom+0x69														/usr/local/go/src/net/udpsock_posix.go:47
#	0x1196fd1	net.(*UDPConn).ReadFrom+0x71														/usr/local/go/src/net/udpsock.go:118
#	0x17e4984	github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/httpu.(*HTTPUClient).Do+0x724				/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/httpu/httpu.go:102
#	0x17e5878	github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/ssdp.SSDPRawSearchCtx+0x588				/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/ssdp/ssdp.go:58
#	0x180788a	github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp.DiscoverDevicesCtx+0x10a					/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/goupnp.go:72
#	0x1808709	github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp.NewServiceClientsCtx+0x69					/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/service_client.go:33
#	0x1809ff1	github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/dcps/internetgateway1.NewWANIPConnection1Clients+0x61	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/goupnp/dcps/internetgateway1/internetgateway1.go:1846
#	0x1811342	github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp.DiscoverCtx+0xc2							/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/NebulousLabs/go-upnp/upnp.go:138
#	0x186212a	main.configureUpnp+0x4a															/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/server.go:126
#	0x18657a2	main.newServer+0x2aa2															/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/server.go:281
#	0x1838d21	main.lndMain+0xa01															/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/lnd.go:239
#	0x183ae75	main.main+0x35																/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/lnd.go:505
#	0x102eb5c	runtime.main+0x20c				
```

**Comment by johng** [comment]
@Roasbeef  what did you change to I've modified the output of the NATPMP print statement, so it should be in the correct human readable format (according to the docs).

I double checked my setup and everything was working with UPNP, external peers were able to successfully connect. What did you change to cause the panic?

**Comment by Roasbeef** [comment]
> What did you change to cause the panic?

That wasn't a panic, it was just a manually extracted gorutine dump. What happened was, after a restart `lnd` got stuck there, and the active goroutine is the one I linked above.

**Comment by halseth** [comment]
Did a quick test of this, and both `upnp` and `natpmp` were able to detect my IP :)

**Comment by halseth** [comment]
Can be worth taking a look at: github.com/NebulousLabs/Sia/pull/2839

**Comment by Roasbeef** [comment]
Superseded by #1109.

**Pull request closed without merging**